rust/kernel/file.rs

```diff
@@ -15,51 +15,57 @@ use crate::{
     sync::CondVar,
     types::PointerWrapper,
     user_ptr::{UserSlicePtr, UserSlicePtrReader, UserSlicePtrWriter},
+    ARef, AlwaysRefCounted,
 };
 use core::convert::{TryFrom, TryInto};
-use core::{marker, mem, mem::ManuallyDrop, ops::Deref, ptr};
+use core::{cell::UnsafeCell, marker, mem, ptr};
 
 /// Wraps the kernel's `struct file`.
-///
-/// # Invariants
-///
-/// The pointer `File::ptr` is non-null and valid. Its reference count is also non-zero.
-pub struct File {
-    pub(crate) ptr: *mut bindings::file,
-}
```

Comment:
Maybe list as invariant that it will never end up on the stack but always as it's own heap allocation? That is necessary for safe refcounting.

Reply:
This is a safety requirement of the `AlwaysRefCounted` trait -- note that it's `unsafe`. Line 82 says that this is the case, so I'm not sure we need to explicitly say it again here as an invariant.

Reply:
Line 82 is admitting (satisfying) the safety condition of the `AlwaysRefCounted` trait using this invariant of the `File` type. While it is possible to infer it based on the `/// SAFETY` comment, I think it would be slightly nicer to explicitly state the invariant on the type. I'm fine either way though.

Reply:
That's a fair point. I suppose if we document this as an invariant, we can use this fact to discharge the obligation we get when implementing `AlwaysRefCounted`.

I'll make this change.